tests/end2end/fixtures/qutewm_sub.py

```diff
@@ -0,0 +1,400 @@
+        self.dpy = Display()
+        self.dpy.set_error_handler(self._error_handler)
+
+        screen_width = self.dpy.screen().width_in_pixels
+        screen_height = self.dpy.screen().height_in_pixels
+        self.dimensions = (screen_width, screen_height)
+
+        self._retcode = None
+        self._needs_update = False
+        self.root = self.dpy.screen().root
+        self.support_window = None
+        self.windows = []
+        self.window_stack = []
+
+        self.root.change_attributes(event_mask=self.ROOT_EVENT_MASK,
+                                    onerror=self._wm_running)
+
+        self.atoms = AtomBag(self.dpy)
+
+        self._set_supported_attribute()
+        self._set_supporting_wm_check()
+
+    def _wm_running(self, error, request):
+        """Called when another WM is already running."""
+        log.error("Another window manager is running, exiting")
+        # This is called async, which means we can't just raise an exception,
+        # we need to signal the main thread to stop.
+        self._retcode = 42
+
+    def _error_handler(self, error, request):
+        # This is called async, which means we can't just raise an exception,
+        # we need to signal the main thread to stop.
+        if self._retcode is None:
+            log.error(error)
+
+    def _set_supported_attribute(self):
+        """Set the _NET_SUPPORTED attribute on the root window."""
+        attributes = [
+            self.atoms.supported,
+            self.atoms.active_window,
+            self.atoms.client_list,
+            self.atoms.wm_state,
+        ]
+        self.root.change_property(
+            self.atoms.supported,
+            Xatom.ATOM,
+            32,
+            attributes,
+        )
+
+    def _set_supporting_wm_check(self):
+        """Create and set a window for _NET_SUPPORTING_WM_CHECK."""
+        self.support_window = self.root.create_window(
+            0, 0, 10, 10, 0, self.dpy.screen().root_depth)
+
+        for window in [self.root, self.support_window]:
+            window.change_property(
+                self.atoms.supporting_wm,
+                Xatom.WINDOW,
+                32,
+                [self.support_window.id],
+            )
+        self.support_window.change_property(
+            self.atoms.net_wm_name,
+            Xatom.STRING,
+            8,
+            self.WM_NAME,
+        )
+
+    def loop(self):
+        """Start the X event loop.
+
+        Return:
+            The manager's exit code.
+        """
+        if self._retcode is not None:
+            # avoid the "event loop started" message if we exit anyway
+            return self._retcode
+        log.info("event loop started")
+        while 1:
+            if self._retcode is not None:
+                return self._retcode
+            try:
+                ev = self.root.display.next_event()
+                handler = self._handlers.get(ev.type)
+                if handler:
+                    handler(ev)
```

Comment:
Probably not, X can generate lots of events, it would probably just clutter the log. At least I'd hide it behing some verbose flag, which you can turn on when you really want it (probably only when you're working on qutewm itself).